File: daehwa_park/1주차/PGS_택배_배달과_수거하기.py
def solution(cap, n, deliveries, pickups):
    answer = 0
    index = n - 1

    for i in range(n - 1, -1, -1):
        index = i
        if deliveries[i] > 0 or pickups[i] > 0:
            break

    curd = 0
    curp = 0
    for i in range(n - 1, -1, -1):
        curd += deliveries[i]
        curp += pickups[i]

        while curd > cap or curp > cap:
            curd -= cap
            curp -= cap
            answer += (index + 1) * 2
            index = i

    if curd != 0 and curp != 0:
        answer += (index + 1) * 2

    return answer

File: sylee723/1주차/PGS_택배_배달과_수거하기.java
class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        int d_max = n;
        int p_max = n;
        int max_dist;
        long answer = 0;
        
        while (true) {   
            while (d_max > 0 && deliveries[d_max - 1] == 0) {
                d_max--;
            }
            while (p_max > 0 && pickups[p_max - 1] == 0) {
                p_max--;
            }
            
            max_dist = Math.max(d_max, p_max);
            
            if (max_dist == 0) {
                break;
            }
            
            answer += (max_dist * 2);
            
            moveBox(deliveries, d_max, cap);
            moveBox(pickups, p_max, cap);
        }
        
        return answer;
    }
    
    static void moveBox(int[] box, int dist, int cap) {
        int count = 0;
        int portion;
        for (int i = dist - 1; i >= 0; i--) {
            if (count + box[i] <= cap) {
                count += box[i];
                box[i] = 0;
            } else {
                portion = cap - count;
                count += portion;
                box[i] -= portion;
            }

            if (count == cap)
                return;
        }
    }
}